File: app/src/main/java/ohi/andre/consolelauncher/managers/termux/TermuxBridgeCache.java
package ohi.andre.consolelauncher.managers.termux;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class TermuxBridgeCache {

    private static final long REQUEST_TTL_MS = 2500;
    private static final int MAX_CACHE_ENTRIES = 64;
    private static final int MAX_REQUEST_ENTRIES = 128;
    private static final Map<String, Entry> dirs = newBoundedMap(MAX_CACHE_ENTRIES);
    private static final Map<String, Entry> files = newBoundedMap(MAX_CACHE_ENTRIES);
    private static final Map<String, Long> requests = newBoundedMap(MAX_REQUEST_ENTRIES);

    private static <T> Map<String, T> newBoundedMap(final int maxEntries) {
        return new LinkedHashMap<String, T>(16, 0.75f, true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry<String, T> eldest) {
                return size() > maxEntries;
            }
        };
    }

    public static synchronized void putDirs(String path, String stdout) {
        dirs.put(path, new Entry(parse(stdout)));
    }

    public static synchronized void putFiles(String path, String stdout) {
        files.put(path, new Entry(parse(stdout)));
    }

    public static synchronized List<String> dirs(String path) {
        Entry entry = dirs.get(path);
        return entry == null ? Collections.emptyList() : entry.values;
    }

    public static synchronized List<String> files(String path) {
        Entry entry = files.get(path);
        return entry == null ? Collections.emptyList() : entry.values;
    }

    public static synchronized boolean shouldRequest(String type, String path) {
        String key = type + ":" + path;
        long now = System.currentTimeMillis();
        Long last = requests.get(key);
        if (last != null && now - last < REQUEST_TTL_MS) {
            return false;
        }
        requests.put(key, now);
        return true;
    }

    private static List<String> parse(String stdout) {
        List<String> values = new ArrayList<>();
        if (stdout == null) {
            return values;
        }

        String[] lines = stdout.trim().split("\\n");
        for (String line : lines) {
            String value = line.trim();
            if (value.endsWith("/")) {
                value = value.substring(0, value.length() - 1);
            }
            if (value.length() > 0) {
                values.add(value);
            }
        }
        Collections.sort(values, String.CASE_INSENSITIVE_ORDER);
        return values;
    }

    private static class Entry {
        final List<String> values;

        Entry(List<String> values) {
            this.values = values;
        }
    }
}

File: app/src/main/java/ohi/andre/consolelauncher/managers/termux/TermuxBridgeCache.kt
package ohi.andre.consolelauncher.managers.termux

import java.util.ArrayList
import java.util.Collections
import java.util.LinkedHashMap
import java.util.Locale
import java.util.Map

object TermuxBridgeCache {
    private const val REQUEST_TTL_MS = 2500L
    private const val MAX_CACHE_ENTRIES = 64
    private const val MAX_REQUEST_ENTRIES = 128
    private val dirs = newBoundedMap<Entry>(MAX_CACHE_ENTRIES)
    private val files = newBoundedMap<Entry>(MAX_CACHE_ENTRIES)
    private val requests = newBoundedMap<Long>(MAX_REQUEST_ENTRIES)

    private fun <T> newBoundedMap(maxEntries: Int): MutableMap<String, T> =
        object : LinkedHashMap<String, T>(16, 0.75f, true) {
            override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, T>?): Boolean = size > maxEntries
        }

    @JvmStatic
    @Synchronized
    fun putDirs(path: String, stdout: String?) {
        dirs[path] = Entry(parse(stdout))
    }

    @JvmStatic
    @Synchronized
    fun putFiles(path: String, stdout: String?) {
        files[path] = Entry(parse(stdout))
    }

    @JvmStatic
    @Synchronized
    fun dirs(path: String): List<String> {
        val entry = dirs[path]
        return entry?.values ?: Collections.emptyList()
    }

    @JvmStatic
    @Synchronized
    fun files(path: String): List<String> {
        val entry = files[path]
        return entry?.values ?: Collections.emptyList()
    }

    @JvmStatic
    @Synchronized
    fun shouldRequest(type: String, path: String): Boolean {
        val key = "$type:$path"
        val now = System.currentTimeMillis()
        val last = requests[key]
        if (last != null && now - last < REQUEST_TTL_MS) {
            return false
        }
        requests[key] = now
        return true
    }

    private fun parse(stdout: String?): List<String> {
        val values: MutableList<String> = ArrayList()
        if (stdout == null) {
            return values
        }

        val lines = stdout.trim().split("\\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (line in lines) {
            var value = line.trim()
            if (value.endsWith("/")) {
                value = value.substring(0, value.length - 1)
            }
            if (value.isNotEmpty()) {
                values.add(value)
            }
        }
        Collections.sort(values, String.CASE_INSENSITIVE_ORDER)
        return values
    }

    private class Entry(val values: List<String>)
}
